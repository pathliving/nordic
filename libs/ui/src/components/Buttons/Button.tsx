'use client';

import clsx from 'clsx';
import { ButtonHTMLAttributes, forwardRef, ReactNode } from 'react';
import style from './button.module.css';

import '@/styles/colors.css';
import '@/styles/font.css';
import '@/styles/spacing.css';
import '@/styles/transition.css';

export type Props = ButtonHTMLAttributes<HTMLButtonElement> & {
  /**
   * What content to add
   */
  children: ReactNode;
  /**
   * What additional styles to add
   */
  className?: string;
  /**
   * What size to use
   */
  size?: 's' | 'm' | 'l';
  /**
   * What background color to use
   */
  appearance?: 'primary' | 'secondary';
  /**
   * What icon to add on the left
   */
  leftIcon?: ReactNode;
  /**
   * What icon to add on the right
   */
  rightIcon?: ReactNode;
  /**
   * What loader to add
   */
  loader?: ReactNode;
  /**
   * What state of loader
   */
  isLoading?: boolean;
};

/**
 * Primary Button UI component for user interaction
 */
const Button = forwardRef<HTMLButtonElement, Props>((props, ref) => {
  const {
    children,
    className,
    size = 'm',
    appearance = 'primary',
    type = 'button',
    leftIcon,
    rightIcon,
    loader = 'wait...',
    isLoading,
    disabled,
    ...rest
  } = props;

  const isDisabled = Boolean(disabled || isLoading);

  return (
    <button
      className={clsx(style.button, style[appearance], style[size], className)}
      type={type}
      disabled={isDisabled}
      {...(isDisabled && {
        'aria-disabled': true,
      })}
      ref={ref}
      {...rest}
    >
      <span className={style.content}>
        {leftIcon && <span aria-hidden="true">{leftIcon}</span>}
        {isLoading ? <span>{loader}</span> : children}
        {rightIcon && <span aria-hidden="true">{rightIcon}</span>}
      </span>
    </button>
  );
});

export default Button;
