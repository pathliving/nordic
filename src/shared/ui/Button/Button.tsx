import { Button as ButtonUI } from '@radix-ui/themes';
import { ComponentPropsWithRef } from 'react';

export default function Button(props: ComponentPropsWithRef<typeof ButtonUI>) {
  return <ButtonUI {...props} />;
}
