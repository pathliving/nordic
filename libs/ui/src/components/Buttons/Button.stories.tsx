import { fn } from "@storybook/test";
import type { Meta, StoryObj } from "@storybook/react";
import { useEffect, useState } from "react";
import Button from "./Button";

const meta: Meta<typeof Button> = {
  title: "Components/Button",
  component: Button,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  args: {
    onClick: fn(),
  },
};
export default meta;

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    children: "Button",
    className: 'btn-primary',
    size: "m",
    appearance: "primary",
    type: "button",
    leftIcon: '',
    rightIcon: '',
    loader: '',
    isLoading: false,
    disabled: false,
  },
};

export const Secondary = {
  args: {
    ...Primary.args,
    appearance: "secondary",
  },
};

export const PrimaryLoading = (args: Partial<typeof Primary.args>) => {
  const [isLoading, setLoading] = useState(args?.isLoading);

  const handleClick = () => {
    setLoading(true);
  };

  useEffect(() => {
    if (!isLoading) return;

    const timeout = setTimeout(() => {
      setLoading(false);
    }, 1500);

    return () => {
      clearTimeout(timeout);
    };
  });

  return (
    <Button onClick={handleClick} loader="wait ..." isLoading={isLoading}>
      Button
    </Button>
  );
};

PrimaryLoading.args = {
  isLoading: true,
};

export const LeftIcon = {
  args: {
    ...Primary.args,
    children: "Activate",
    leftIcon: "⚡️",
  },
};

export const RightIcon = {
  args: {
    ...Primary.args,
    children: "Deactivate",
    rightIcon: "⚡️",
  },
};

export const Small = {
  args: {
    ...Primary.args,
    size: "s",
  },
};

export const Medium = {
  args: {
    ...Primary.args,
    size: "m",
  },
};

export const Large = {
  args: {
    ...Primary.args,
    size: "l",
  },
};
