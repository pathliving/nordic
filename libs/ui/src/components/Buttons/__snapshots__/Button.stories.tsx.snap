// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Button Large smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _l_cvwg2_45 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Button
  </span>
</button>
`;

exports[`Components/Button LeftIcon smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _m_cvwg2_35 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    <span aria-hidden="true">
      ⚡️
    </span>
    Activate
  </span>
</button>
`;

exports[`Components/Button Medium smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _m_cvwg2_35 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Button
  </span>
</button>
`;

exports[`Components/Button Primary smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _m_cvwg2_35 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Button
  </span>
</button>
`;

exports[`Components/Button PrimaryLoading smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _m_cvwg2_35"
        type="button"
        disabled
        aria-disabled="true"
>
  <span class="_content_cvwg2_17">
    <span>
      wait ...
    </span>
  </span>
</button>
`;

exports[`Components/Button RightIcon smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _m_cvwg2_35 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Deactivate
    <span aria-hidden="true">
      ⚡️
    </span>
  </span>
</button>
`;

exports[`Components/Button Secondary smoke-test 1`] = `
<button class="_button_cvwg2_1 _secondary_cvwg2_70 _m_cvwg2_35 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Button
  </span>
</button>
`;

exports[`Components/Button Small smoke-test 1`] = `
<button class="_button_cvwg2_1 _primary_cvwg2_55 _s_cvwg2_25 btn-primary"
        type="button"
>
  <span class="_content_cvwg2_17">
    Button
  </span>
</button>
`;
